db/crud.go

```diff
@@ -679,54 +657,86 @@ func (db *Database) Put(docid string, body Body) (newRevID string, doc *Document
 		if doc != nil && !isSgWrite && db.UseXattrs() {
-			err := db.OnDemandImportForWrite(newDoc.ID, doc, deleted)
+			err := db.OnDemandImportForWrite(newDoc.ID, doc, newDoc.Deleted)
 			if err != nil {
 				return nil, nil, nil, err
 			}
 		}
 
 		// First, make sure matchRev matches an existing leaf revision:
-		if matchRev == "" {
-			matchRev = doc.CurrentRev
-			if matchRev != "" {
+		if newDoc.RevID == "" {
+			newDoc.RevID = doc.CurrentRev
+			if newDoc.RevID != "" {
 				// PUT with no parent rev given, but there is an existing current revision.
 				// This is OK as long as the current one is deleted.
-				if !doc.History[matchRev].Deleted {
+				if !doc.History[newDoc.RevID].Deleted {
 					return nil, nil, nil, base.HTTPErrorf(http.StatusConflict, "Document exists")
 				}
-				generation, _ = ParseRevID(matchRev)
+				generation, _ = ParseRevID(newDoc.RevID)
 				generation++
 			}
-		} else if !doc.History.isLeaf(matchRev) || db.IsIllegalConflict(doc, matchRev, deleted, false) {
+		} else if !doc.History.isLeaf(newDoc.RevID) || db.IsIllegalConflict(doc, newDoc.RevID, newDoc.Deleted, false) {
 			return nil, nil, nil, base.HTTPErrorf(http.StatusConflict, "Document revision conflict")
 		}
 
 		// Process the attachments, and populate _sync with metadata. This alters 'body' so it has to
 		// be done before calling createRevID (the ID is based on the digest of the body.)
-		newAttachments, err := db.storeAttachments(doc, newDoc.DocAttachments, generation, matchRev, nil)
+		newAttachments, err := db.storeAttachments(doc, newDoc.DocAttachments, generation, newDoc.RevID, nil)
 		if err != nil {
 			return nil, nil, nil, err
 		}
 
 		// Make up a new _rev, and add it to the history:
-		newRev, err := createRevID(generation, matchRev, newDoc.Body())
+		newRev, err := createRevID(generation, newDoc.RevID, newDoc.Body())
 		if err != nil {
 			return nil, nil, nil, err
 		}
 
-		if err := doc.History.addRevision(newDoc.ID, RevInfo{ID: newRev, Parent: matchRev, Deleted: deleted}); err != nil {
-			base.InfofCtx(db.Ctx, base.KeyCRUD, "Failed to add revision ID: %s, for doc: %s, error: %v", newRev, base.UD(docid), err)
+		if err := doc.History.addRevision(newDoc.ID, RevInfo{ID: newRev, Parent: newDoc.RevID, Deleted: newDoc.Deleted}); err != nil {
+			base.InfofCtx(db.Ctx, base.KeyCRUD, "Failed to add revision ID: %s, for doc: %s, error: %v", newRev, base.UD(newDoc.ID), err)
 			return nil, nil, nil, base.ErrRevTreeAddRevFailure
 		}
 
 		// move _attachment metadata to syncdata of doc after rev-id generation
 		doc.SyncData.Attachments = newDoc.DocAttachments
 		newDoc.RevID = newRev
-		newDoc.Deleted = deleted
 
 		return newDoc, newAttachments, nil, nil
 	})
 
+	return
+
+}
+
+// Updates or creates a document.
+// The new body's BodyRev property must match the current revision's, if any.
+func (db *Database) PutWithBody(docid string, body Body) (newRevID string, doc *Document, err error) {
```

Comment:
Since the replicator2 path is going straight to PutExistingRev, I expected we could remove the Put/PutWithBody changes.